### PR TITLE
fix: integrate with `allocate` API for creating assignments, update global retry behavior on queries, handle 404 subsidy access policy

### PR DESCRIPTION
**Pull request by adamstankiewicz**
# Description

First of at least two PRs for [ENT-7826](https://2u-internal.atlassian.net/browse/ENT-7826)

Includes:
* Integrates with the `allocate` POST API endpoint in enterprise-access. **Happy path only**, using whatever email(s) are input in the assignment modal.
* Modifies the global retry behavior for `@tanstack/react-query` to avoid retries when the query has thrown max 3 errors in a row (default behavior) or returns a 404 once.
* Displays skeleton loading state when doing background refetches of the budget detail activity overview data (spent table + assigned table, if applicable).
  * This is necessary for when an admin creates their first assignment(s) when the "Activity" tab previously displayed "No budget activity?" empty state such that the empty state doesn't briefly show immediately after creating new assignment(s) while the activity overview data refetches. The budget detail activity overview data refetches after an assignment was made due to invalidating the query cache for the overview data query, e.g.:

```jsx
queryClient.invalidateQueries({
  queryKey: learnerCreditManagementQueryKeys.budgetActivityOverview(subsidyAccessPolicyId),
});
```

* Ensures a "Page Not Found" screen is displayed if the budget detail page route includes an identifier for a subsidy access policy (i.e., a UUID), the data is not currently loading, and is in an error status.
* Ensures the initial loading of subsidy access policy metadata, if any, on the budget detail page does _not_ include the `Hero` component for improved handling of the above "Page Not Found" behavior.

### Creating new assignments

https://github.com/openedx/frontend-app-admin-portal/assets/2828721/2637379d-5d29-42de-a63a-714d42cb4e0e

### Displaying Page Not Found when attempting to load data about a subsidy access policy that doesn't exist (404)

Instead of showing an indefinite loading spinner, the UI will now display "page not found" messaging, since the requested subsidy access policy does not exist.

image
 

# For all changes

- [x] Ensure adequate tests are in place (or reviewed existing tests cover changes)

# Only if submitting a visual change

- [x] Ensure to attach screenshots
- [ ] Ensure to have UX team confirm screenshots

### CODE REVIEW

File: src/utils.js [hunk 1]
```diff
@@ -400,6 +400,18 @@ const pollAsync = async (pollFunc, timeout, interval, checkFunc) => {
   return false;
 };
 
+/**
+ * Modifies the retry behavior of queries to retry up to max 3 times (default) or if
+ * if the error returned by the query is a 404 HTTP status code (not found). This configuration
```
Comment (iloveagent57) [inline src/utils.js:405]: nit: double "if"

File: src/components/learner-credit-management/BudgetDetailPage.jsx [hunk 2]
```diff
@@ -25,6 +28,12 @@ const BudgetDetailPage = () => {
     );
   }
 
+  // If the budget is intended to be a subsidy access policy (by presence of a policy UUID),
+  // and the subsidy access policy is not found, show 404 messaging.
+  if (subsidyAccessPolicyId && isSubsidyAccessPolicyError && error?.customAttributes.httpErrorStatus === 404) {
+    return <NotFoundPage />;
+  }
+
```
Comment (iloveagent57) [inline src/components/learner-credit-management/BudgetDetailPage.jsx:36]: 👍

File: src/data/services/tests/EnterpriseAccessApiService.test.js [hunk 3]
```diff
@@ -152,4 +152,15 @@ describe('EnterpriseAccessApiService', () => {
       `${enterpriseAccessBaseUrl}/api/v1/subsidy-access-policies/${mockSubsidyAccessPolicyUUID}/`,
     );
   });
+
+  test('allocateContentAssignments calls enterprise-access allocate POST API to create assignments', () => {
+    const payload = {
+      learner_emails: ['edx@example.com'],
```
Comment (iloveagent57) [inline src/data/services/tests/EnterpriseAccessApiService.test.js:158]: Nit: is it worth having a test payload that matches what the endpoint actually requires (i.e. a `content_key` and `content_price_cents`)?

**Review by iloveagent57 — APPROVED** [review]

File: src/data/services/tests/EnterpriseAccessApiService.test.js (hunk 3 above)
Comment (adamstankiewicz) [reply src/data/services/tests/EnterpriseAccessApiService.test.js:158]: Updated!

**Pull request merged**